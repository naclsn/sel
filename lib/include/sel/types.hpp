#ifndef SEL_TYPE_HPP
#define SEL_TYPE_HPP

/**
 * Parsing and representation of types, as well as a set
 * of helper functions to deal with the type system.
 */

#include <ostream>
#include <istream>
#include <string>
#include <vector>
#include <iterator>


namespace sel {

  enum class Ty {
    UNK,
    NUM,
    STR,
    LST,
    FUN,
  };
  enum TyFlag {
    IS_FIN = 0, // (0, and as such default)
    IS_INF = 1,
    IS_TPL = 2,
  };

  /**
   * Represents a type (yey). Use the [..]Type function
   * to construct/parse.
   */
  struct Type {
    Ty base = Ty::UNK;
    union P {
      std::string* name;
      std::vector<Type*>* box_has; // Ty has;
      Type* box_pair[2]; // Ty pair[2];
    } p = {.name=nullptr};
    uint8_t flags = 0;

    Type() { }
    Type(Ty base, Type::P p, uint8_t flags);
    Type(Type const& ty);
    Type(Type&& ty) noexcept;
    ~Type();

    bool operator==(Type const& other) const;
    bool operator!=(Type const& other) const;

    std::vector<Type*>& has() const { return *p.box_has; }
    Type const& from() const { return *p.box_pair[0]; }
    Type const& to() const { return *p.box_pair[1]; }
    Type applied(Type const& arg) const;
  };

  /**
   * Parse a type from the given stream. The overload to
   * the `>>` operator is also provided as a convenience,
   * but this function allows retreiving a name
   * (which would be specified with the `<name> ::`
   * syntax). If `named` is not nullptr but the parsed
   * type representation does not specify a name, the
   * empty string is written into `named`.
  */
  void parseType(std::istream& in, std::string* named, Type& res);

  std::ostream& operator<<(std::ostream& out, Type const& ty);
  std::istream& operator>>(std::istream& in, Type& res);

} // namespace sel

#endif // SEL_TYPE_HPP
