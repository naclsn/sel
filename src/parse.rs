use std::io;
use std::iter;

#[derive(PartialEq, Debug)]
pub enum Token {
    Unknown(String),
    Word(String),
    Text(Vec<u8>),
    Number(i32),
    Comma,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
}

struct Bytes<T: io::Read> {
    inner: io::Bytes<T>,
}

impl<T: io::Read> Iterator for Bytes<T> {
    type Item = u8;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next().and_then(|e| e.ok())
    }
}

pub(crate) struct Lexer<T: io::Read> {
    bytes: iter::Peekable<Bytes<T>>,
}

impl<T: io::Read> Lexer<T> {
    fn new(stream: T) -> Self {
        Self {
            bytes: Bytes {
                inner: stream.bytes(),
            }
            .peekable(),
        }
    }
}

impl<T: io::Read> Iterator for Lexer<T> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        Some(match self.bytes.find(|c| !c.is_ascii_whitespace())? {
            b':' => Token::Text({
                iter::from_fn(|| match (self.bytes.next(), self.bytes.peek()) {
                    (Some(b':'), Some(b':')) => {
                        let _ = self.bytes.next();
                        Some(b':')
                    }
                    (Some(b':'), _) => None,
                    (c, _) => c,
                })
                .collect()
            }),

            b',' => Token::Comma,
            b'[' => Token::OpenBracket,
            b']' => Token::CloseBracket,
            b'{' => Token::OpenBrace,
            b'}' => Token::CloseBrace,

            b'#' => self.bytes.find(|&c| b'\n' == c).and_then(|_| self.next())?,

            c if c.is_ascii_lowercase() => Token::Word(
                String::from_utf8(
                    iter::once(c)
                        .chain(iter::from_fn(|| self.bytes.next_if(u8::is_ascii_lowercase)))
                        .collect(),
                )
                .unwrap(),
            ),

            c if c.is_ascii_digit() => Token::Number({
                let mut r = 0i32;
                let (shift, digits) = match (c, self.bytes.peek()) {
                    (b'0', Some(b'B' | b'b')) => {
                        self.bytes.next();
                        (1, b"01".as_slice())
                    }
                    (b'0', Some(b'O' | b'o')) => {
                        self.bytes.next();
                        (3, b"01234567".as_slice())
                    }
                    (b'0', Some(b'X' | b'x')) => {
                        self.bytes.next();
                        (4, b"0123456789abcdef".as_slice())
                    }
                    _ => {
                        r = c as i32 - 48;
                        (0, b"0123456789".as_slice())
                    }
                };
                while let Some(k) = self
                    .bytes
                    .peek()
                    .and_then(|&c| digits.iter().position(|&k| k == c | 32))
                {
                    self.bytes.next();
                    r = if 0 == shift {
                        r * 10 + k as i32
                    } else {
                        r << shift | k as i32
                    }
                }
                r
            }),

            c => Token::Unknown(
                String::from_utf8_lossy(
                    &iter::once(c)
                        .chain(self.bytes.by_ref().take_while(|c| !c.is_ascii_whitespace()))
                        .collect::<Vec<_>>(),
                )
                .into_owned(),
            ),
        })
    }
}

#[derive(PartialEq, Debug)]
pub enum Tree {
    Apply(Box<Tree>, Vec<Tree>),
    Atom(Token),
    Chain(Vec<Tree>),
    List(Vec<Tree>),
}

#[derive(PartialEq, Debug)]
pub enum Error {
    Unexpected(Token),
    UnexpectedEOF,
}

impl Tree {
    fn one_from_peekable(
        peekable: &mut iter::Peekable<Lexer<impl io::Read>>,
    ) -> Result<Tree, Error> {
        let one = match peekable.next() {
            Some(token @ (Token::Word(_) | Token::Text(_) | Token::Number(_))) => Tree::Atom(token),

            Some(Token::OpenBracket) => Tree::Chain(
                iter::once(Tree::one_from_peekable(peekable))
                    .chain(iter::from_fn(|| match peekable.next() {
                        Some(Token::Comma) => Some(Tree::one_from_peekable(peekable)),
                        Some(Token::CloseBracket) => None,
                        Some(token) => Some(Err(Error::Unexpected(token))),
                        None => Some(Err(Error::UnexpectedEOF)),
                    }))
                    .collect::<Result<_, _>>()?,
            ),

            Some(Token::OpenBrace) => Tree::List(
                iter::once(Tree::one_from_peekable(peekable))
                    .chain(iter::from_fn(|| match peekable.next() {
                        Some(Token::Comma) => Some(Tree::one_from_peekable(peekable)),
                        Some(Token::CloseBrace) => None,
                        Some(token) => Some(Err(Error::Unexpected(token))),
                        None => Some(Err(Error::UnexpectedEOF)),
                    }))
                    .collect::<Result<_, _>>()?,
            ),

            Some(token) => return Err(Error::Unexpected(token)),
            None => return Err(Error::UnexpectedEOF),
        };

        let args: Vec<_> = iter::from_fn(|| match peekable.peek() {
            Some(Token::Comma | Token::CloseBracket | Token::CloseBrace) | None => None,
            _ => Some(Tree::one_from_peekable(peekable)),
        })
        .collect::<Result<_, _>>()?;
        Ok(if args.is_empty() {
            one
        } else {
            Tree::Apply(Box::new(one), args)
        })
    }

    fn from_peekable(peekable: &mut iter::Peekable<Lexer<impl io::Read>>) -> Result<Tree, Error> {
        let mut r = vec![Tree::one_from_peekable(peekable)?];
        while let Some(Token::Comma) = peekable.next() {
            r.push(Tree::one_from_peekable(peekable)?);
        }
        Ok(Tree::Chain(r))
    }

    pub fn from_lexer(lexer: Lexer<impl io::Read>) -> Result<Tree, Error> {
        Tree::from_peekable(&mut lexer.peekable())
    }

    pub fn from_stream(stream: impl io::Read) -> Result<Tree, Error> {
        Tree::from_lexer(Lexer::new(stream))
    }

    pub fn from_str(text: &str) -> Result<Tree, Error> {
        Tree::from_stream(text.as_bytes())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lexing() {
        assert_eq!(
            &[Token::Word("coucou".into()),][..],
            Lexer::new("coucou".as_bytes()).collect::<Vec<Token>>()
        );

        assert_eq!(
            &[
                Token::OpenBracket,
                Token::Word("a".to_string()),
                Token::Number(1),
                Token::Word("b".to_string()),
                Token::Comma,
                Token::OpenBrace,
                Token::Word("w".to_string()),
                Token::Comma,
                Token::Word("x".to_string()),
                Token::Comma,
                Token::Word("y".to_string()),
                Token::Comma,
                Token::Word("z".to_string()),
                Token::CloseBrace,
                Token::CloseBracket,
            ][..],
            Lexer::new("[a 1 b, {w, x, y, z}]".as_bytes()).collect::<Vec<Token>>()
        );

        assert_eq!(
            &[
                Token::Number(42),
                Token::Text(vec![42]),
                Token::Number(42),
                Token::Number(42),
                Token::Number(42),
            ][..],
            Lexer::new("42 :*: 0x2a 0b101010 0o52".as_bytes()).collect::<Vec<Token>>()
        );
    }

    #[test]
    fn parsing() {
        assert_eq!(
            Ok(Tree::Chain(vec![Tree::Atom(Token::Word("coucou".to_string()))])),
            Tree::from_str("coucou")
        );

        assert_eq!(
            Ok(Tree::Chain(vec![
                Tree::Apply(
                    Box::new(Tree::Atom(Token::Word("a".to_string()))),
                    vec![Tree::Apply(
                        Box::new(Tree::Atom(Token::Number(1))),
                        vec![Tree::Atom(Token::Word("b".to_string()))],
                    )],
                ),
                Tree::List(vec![
                    Tree::Atom(Token::Word("w".to_string())),
                    Tree::Atom(Token::Word("x".to_string())),
                    Tree::Atom(Token::Word("y".to_string())),
                    Tree::Atom(Token::Word("z".to_string())),
                ])
            ])),
            Tree::from_str("a 1 b, {w, x, y, z}")
        );

        // todo
        //assert_eq!(Err...)
    }
}
