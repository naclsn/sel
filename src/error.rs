use std::fmt::{Display, Formatter, Result as FmtResult};

use crate::parse::{Applicable, TokenKind};
use crate::scope::{Location, SourceRegistry};
use crate::types::FrozenType;

// error types {{{
#[derive(PartialEq, Debug)]
pub enum ErrorContext {
    Unmatched {
        open_token: TokenKind,
    },
    CompleteType {
        complete_type: FrozenType,
    },
    TypeListInferredItemType {
        list_item_type: FrozenType,
    },
    AsNthArgToNowTyped {
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromAsNthArgToNowTyped {
        comma_loc: Location,
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromToNotFunc {
        comma_loc: Location,
    },
    AutoCoercedVia {
        func_name: String,
        func_type: FrozenType,
    },
    DeclaredHereWithType {
        with_type: FrozenType,
    },
    LetFallbackTypeMismatch {
        result_type: FrozenType,
        fallback_type: FrozenType,
    },
}

#[derive(PartialEq, Debug)]
pub enum ErrorKind {
    ContextCaused {
        error: Box<Error>,
        because: ErrorContext,
    },
    Unexpected {
        token: TokenKind,
        expected: &'static str,
    },
    UnexpectedDefInScript,
    UnknownName {
        name: String,
        expected_type: FrozenType,
    },
    NotFunc {
        actual_type: FrozenType,
    },
    TooManyArgs {
        nth_arg: usize,
        func: Applicable,
    },
    ExpectedButGot {
        expected: FrozenType,
        actual: FrozenType,
    },
    InfWhereFinExpected,
    NameAlreadyDeclared {
        name: String,
    },
    CouldNotReadFile {
        error: String,
    },
}

#[derive(PartialEq, Debug)]
pub struct Error(pub Location, pub ErrorKind);

#[derive(Default, PartialEq, Debug)]
pub struct ErrorList(pub(crate) Vec<Error>);

impl IntoIterator for ErrorList {
    type Item = Error;
    type IntoIter = <Vec<Error> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl ErrorList {
    pub(crate) fn push(&mut self, err: Error) {
        self.0.push(err);
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}
// }}}

pub struct Report<'a> {
    registry: &'a SourceRegistry,
    title: String,
    messages: Vec<(Location, String)>,
}

// generate report {{{
impl Error {
    fn ctx_messages(loc: Location, because: &ErrorContext, report: &mut Report) {
        use ErrorContext::*;
        use TokenKind::*;

        let msgs: &[_] = match because {
            Unmatched { open_token } => &[(
                loc,
                format!(
                    "open {} here",
                    match open_token {
                        OpenBracket => '[',
                        OpenBrace => '{',
                        Unknown(_) | Word(_) | Bytes(_) | Number(_) | Comma | CloseBracket
                        | CloseBrace | Equal | Def | Let | Use | Semicolon | End => unreachable!(),
                    }
                ),
            )],
            CompleteType { complete_type } => &[(loc, format!("complete type: {complete_type}"))],
            TypeListInferredItemType { list_item_type } => &[(
                loc,
                format!("list type was inferred to be [{list_item_type}]"),
            )],
            AsNthArgToNowTyped {
                nth_arg,
                func,
                type_with_curr_args,
            } => &[(
                loc,
                format!(
                    "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                    match nth_arg {
                        1 => "st",
                        2 => "nd",
                        3 => "rd",
                        _ => "th",
                    },
                    match func {
                        Applicable::Name(name) => name,
                        Applicable::Bind(_, _, _) => "let binding",
                    },
                ),
            )],
            ChainedFromAsNthArgToNowTyped {
                comma_loc,
                nth_arg,
                func,
                type_with_curr_args,
            } => &[
                (
                    loc,
                    format!(
                        "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                        match nth_arg {
                            1 => "st",
                            2 => "nd",
                            3 => "rd",
                            _ => "th",
                        },
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                    ),
                ),
                (comma_loc.clone(), "chained through here".into()),
            ],
            ChainedFromToNotFunc { comma_loc } => &[
                (loc, "Not a function".into()),
                (comma_loc.clone(), "chained through here".into()),
            ],
            AutoCoercedVia {
                func_name,
                func_type,
            } => &[(
                loc,
                format!("coerced via '{func_name}' (which has type {func_type})"),
            )],
            DeclaredHereWithType { with_type } => &[(loc, format!("here with type {with_type}"))],
            LetFallbackTypeMismatch {
                result_type,
                fallback_type,
            } => &[(
                loc,
                format!("fallback of type {fallback_type} doesn't match result type {result_type}"),
            )],
        };
        report.messages.extend_from_slice(msgs);
    }

    pub fn report<'a>(&self, registry: &'a SourceRegistry) -> Report<'a> {
        use ErrorKind::*;
        use TokenKind::*;

        let loc = self.0.clone();

        match &self.1 {
            ContextCaused { error, because } => {
                let mut r = error.report(registry);
                Error::ctx_messages(loc, because, &mut r);
                r
            }
            Unexpected { token, expected } => Report {
                registry,
                title: "Unexpected token".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Unexpected '{}', expected {expected}",
                        match token {
                            Unknown(t) => t,
                            Word(v) => v,
                            Bytes(_) => "(bytes)",
                            Number(_) => "(number)",
                            Comma => ",",
                            OpenBracket => "[",
                            CloseBracket => "]",
                            OpenBrace => "{",
                            CloseBrace => "}",
                            Equal => "=",
                            Def => "def",
                            Let => "let",
                            Use => "use",
                            Semicolon => ";",
                            End => "end of script",
                        }
                    ),
                )],
            },
            UnexpectedDefInScript => Report {
                registry,
                title: "Unexpected definition".into(),
                messages: vec![(
                    loc,
                    "these should only appear before the script, at the beginning of the file"
                        .into(),
                )],
            },
            UnknownName {
                name,
                expected_type,
            } => Report {
                registry,
                title: "Unknown name".into(),
                messages: vec![(
                    loc,
                    format!("Unknown name '{name}', should be of type {expected_type}"),
                )],
            },
            NotFunc { actual_type } => Report {
                registry,
                title: "Not a function".into(),
                messages: vec![(
                    loc,
                    format!("Expected a function type, but got {actual_type}"),
                )],
            },
            TooManyArgs { nth_arg, func } => Report {
                registry,
                title: "Too many argument".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Too many argument to {}, expected only {}",
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                        nth_arg - 1
                    ),
                )],
            },
            ExpectedButGot { expected, actual } => Report {
                registry,
                title: "Type mismatch".into(),
                messages: vec![(loc, format!("Expected type {expected}, but got {actual}"))],
            },
            InfWhereFinExpected => Report {
                registry,
                title: "Wrong boundedness".into(),
                messages: vec![(loc, "Expected finite type, but got infinite type".into())],
            },
            NameAlreadyDeclared { name } => Report {
                registry,
                title: "Name used twice".into(),
                messages: vec![(loc, format!("Name {name} was already declared"))],
            },
            CouldNotReadFile { error } => Report {
                registry,
                title: "Could not read file".into(),
                messages: vec![(loc, format!("<< {error} >>"))],
            },
        }
    }
}
// }}}

// display report {{{
impl Display for Report<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        let (Location(file, ref range), _) = self.messages[0];
        let path = self.registry.get_path(file);
        let bytes = self.registry.get_bytes(file);
        let line = 1 + bytes[range.start..].iter().filter(|c| b'\n' == **c).count();
        writeln!(f, "{}:{line}: {}", path.display(), self.title)?;

        for (_, msg) in &self.messages {
            writeln!(f, " -> {}", msg)?;
        }
        Ok(())
    }
}
// }}}
