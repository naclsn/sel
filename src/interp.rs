use std::cell::RefCell;
use std::io::{self, Read, Write};
use std::iter;
use std::mem;
use std::ptr;
use std::rc::Rc;

use crate::parse::{Tree, TreeLeaf};

pub type Number = Box<dyn FnOnce() -> i32>;
pub type Bytes = Box<dyn Iterator<Item = u8>>;
pub type List = Box<dyn Iterator<Item = Value>>;
pub type Func = Box<dyn FnOnce(Value) -> Value>;

type ValueClone<T> = Rc<dyn Fn() -> T>;

pub enum Value {
    Number(Number, ValueClone<Number>),
    Bytes(Bytes, ValueClone<Bytes>),
    List(List, ValueClone<List>),
    Func(Func, ValueClone<Func>),
}

impl Clone for Value {
    fn clone(&self) -> Self {
        match self {
            Value::Number(_, clone) => {
                let niw = clone();
                Value::Number(niw, clone.clone())
            }
            Value::Bytes(_, clone) => {
                let niw = clone();
                Value::Bytes(niw, clone.clone())
            }
            Value::List(_, clone) => {
                let niw = clone();
                Value::List(niw, clone.clone())
            }
            Value::Func(_, clone) => {
                let niw = clone();
                Value::Func(niw, clone.clone())
            }
        }
    }
}

impl Value {
    fn number(self) -> (Number, ValueClone<Number>) {
        if let Value::Number(r, c) = self {
            (r, c)
        } else {
            unreachable!()
        }
    }
    fn bytes(self) -> (Bytes, ValueClone<Bytes>) {
        if let Value::Bytes(r, c) = self {
            (r, c)
        } else {
            unreachable!()
        }
    }
    fn list(self) -> (List, ValueClone<List>) {
        if let Value::List(r, c) = self {
            (r, c)
        } else {
            unreachable!()
        }
    }
    fn func(self) -> (Func, ValueClone<Func>) {
        if let Value::Func(r, c) = self {
            (r, c)
        } else {
            unreachable!()
        }
    }
}

enum Bival<A, B> {
    Init(A),
    Fini(B),
}

impl<A, B> Bival<A, B> {
    fn new(init: A) -> Bival<A, B> {
        Bival::Init(init)
    }

    fn map(&mut self, f: impl FnOnce(A) -> B) -> &mut B {
        match self {
            Bival::Init(a) => {
                unsafe {
                    let niw = Bival::Fini(f(ptr::read(a)));
                    let Bival::Init(old) = mem::replace(self, niw) else {
                        unreachable!()
                    };
                    _ = mem::MaybeUninit::new(old);
                }
                let Bival::Fini(b) = self else { unreachable!() };
                b
            }
            Bival::Fini(b) => b,
        }
    }
}

fn lookup_val(name: &str, mut args: impl Iterator<Item = Value>) -> Value {
    match name {
        "(,)" => {
            let f = args.next().unwrap().func().0;
            let g = args.next().unwrap().func().0;
            Value::Func(Box::new(move |v| g(f(v))), Rc::new(|| todo!()))
        }

        "input" => {
            let mut stdin = io::stdin();
            Value::Bytes(
                Box::new(iter::from_fn(move || {
                    let mut r = [0u8; 1];
                    match stdin.read(&mut r) {
                        Ok(1) => Some(r[0]),
                        _ => None,
                    }
                })),
                Rc::new(|| todo!()),
            )
        }

        "ln" => Value::Bytes(
            Box::new(args.next().unwrap().bytes().0.chain(iter::once(b'\n'))),
            Rc::new(|| todo!()),
        ),

        "split" => {
            let mut uncollected_sep = Some(args.next().unwrap().bytes().0);
            let mut bytes_iter = args.next().unwrap().bytes().0.peekable();
            let mut sep = Vec::<u8>::new();
            Value::List(
                Box::new(iter::from_fn(move || {
                    bytes_iter.peek()?;
                    if let Some(a) = uncollected_sep.take() {
                        sep = a.collect();
                    }
                    let v: Vec<u8> = bytes_iter.by_ref().take_while(|x| *x != sep[0]).collect();
                    Some(Value::Bytes(Box::new(v.into_iter()), Rc::new(|| todo!())))
                })),
                Rc::new(|| todo!()),
            )
        }

        "add" => {
            let a = args.next().unwrap().number().0;
            if let Some(b) = args.next() {
                let b = b.number().0;
                Value::Number(Box::new(move || a() + b()), Rc::new(|| todo!()))
            } else {
                let a = Rc::new(RefCell::new(Bival::new(a)));
                let f = move |b: Value| {
                    let b = b.number().0;
                    Value::Number(
                        Box::new(move || *a.borrow_mut().map(|a| a()) + b()),
                        Rc::new(|| todo!()),
                    )
                };
                let ff = f.clone();
                Value::Func(Box::new(ff), Rc::new(move || Box::new(f.clone())))
            }
        }

        "map" => {
            let (mut f, f_clone) = args.next().unwrap().func();
            let mut l = args.next().unwrap().list().0;
            Value::List(
                Box::new(iter::from_fn(move || {
                    l.next().map(mem::replace(&mut f, f_clone()))
                })),
                Rc::new(|| todo!()),
            )
        }

        "tonum" => {
            let s = args.next().unwrap().bytes().0;
            Value::Number(
                Box::new(move || {
                    let mut r = 0;
                    for n in s {
                        if n.is_ascii_digit() {
                            r = 10 * r + (n - b'0') as i32;
                        } else {
                            break;
                        }
                    }
                    r
                }),
                Rc::new(|| todo!()),
            )
        }

        "tostr" => {
            let mut n = Bival::new(args.next().unwrap().number().0);
            Value::Bytes(
                Box::new(iter::from_fn(move || {
                    n.map(|n| n().to_string().into_bytes().into_iter()).next()
                })),
                Rc::new(|| todo!()),
            )
        }

        "join" => {
            let mut uncollected_sep = Some(args.next().unwrap().bytes().0);
            let mut list_iter = args.next().unwrap().list().0.map(Value::bytes).peekable();
            let mut sep = Vec::<u8>::new();
            Value::Bytes(
                Box::new(iter::from_fn(move || {
                    if let Some((ref mut bytes, _)) = list_iter.peek_mut() {
                        let b = bytes.next();
                        if b.is_some() {
                            return b;
                        }
                        drop(list_iter.next());
                        if list_iter.peek().is_some() {
                            if let Some(a) = uncollected_sep.take() {
                                sep = a.collect();
                            }
                            return Some(sep[0]);
                        }
                    }
                    None
                })),
                Rc::new(|| todo!()),
            )
        }

        _ => unreachable!(),
    }
}

pub fn interp(tree: &Tree) -> Value {
    match tree {
        Tree::Atom(atom) => match atom {
            TreeLeaf::Bytes(v) => {
                Value::Bytes(Box::new(v.clone().into_iter()), Rc::new(|| todo!()))
            }
            &TreeLeaf::Number(n) => Value::Number(Box::new(move || n), Rc::new(|| todo!())),
        },

        Tree::List(_) => todo!(),

        Tree::Apply(_, name, args) => lookup_val(name, args.iter().map(interp)),
    }
}

pub fn run_print(val: Value) {
    match val {
        Value::Number(n, _) => println!("{}", n()),
        Value::Bytes(b, _) => {
            let mut stdout = io::stdout();
            for ch in b {
                stdout.write_all(&[ch]).unwrap();
            }
        }
        Value::List(l, _) => {
            for it in l {
                run_print(it);
                println!();
            }
        }
        Value::Func(_f, _) => todo!(),
    }
}
