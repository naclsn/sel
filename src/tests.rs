use insta::{assert_debug_snapshot, assert_snapshot};

use crate::error::ErrorList;
use crate::interp;
use crate::parse::{self, Lexer, TokenKind, Tree};
use crate::scope::Global;
use crate::types::FrozenType;

#[test]
fn lexing() {
    fn t(script: &[u8]) -> String {
        let mut end = true;
        Lexer::new(0, script.iter().copied())
            .take_while(|t| {
                if !end {
                    false
                } else {
                    end = TokenKind::End != t.1;
                    true
                }
            })
            .fold(String::new(), |mut acc, cur| {
                acc += &format!("{cur:?}\n");
                acc
            })
    }

    assert_snapshot!(t(b""));
    assert_snapshot!(t(b"coucou"));
    assert_snapshot!(t(b"[a 1 b, {w, x, y, z}, 0.5]"));
    assert_snapshot!(t(b":hay: :hey:: not hay: :: :::: fin"));
    assert_snapshot!(t(b"42 :*: 0x2a 0b101010 0o52"));
    assert_snapshot!(t(b"oh-no; don't let use be the def of cat!"));
}

#[test]
fn parsing() {
    fn t(script: &[u8]) -> Result<(FrozenType, Tree), ErrorList> {
        let mut global = Global::with_builtin();
        let source = global.registry.add_bytes("<test>", script.into());
        let result = parse::process(source, &mut global);
        result
            .tree
            .map(|t| (global.types.frozen(t.ty), t))
            .ok_or(result.errors)
    }

    assert_debug_snapshot!(t(b""));
    assert_debug_snapshot!(t(b"-"));
    assert_debug_snapshot!(t(b"-, split:-:, map[tonum, add1, tostr], join:+:"));
    assert_debug_snapshot!(t(b"tonum, add234121, tostr, ln"));
    assert_debug_snapshot!(t(b"[tonum, add234121, tostr] :13242:"));
    assert_debug_snapshot!(t(b"{repeat 1, {}}"));
    assert_debug_snapshot!(t(b"add 1, tostr"));
    assert_debug_snapshot!(t(b"zipwith add {1}"));
    assert_debug_snapshot!(t(b"{repeat 1, {}}"));
    assert_debug_snapshot!(t(b"{{}, repeat 1}"));
    assert_debug_snapshot!(t(b"{{42}, repeat 1}"));
    assert_debug_snapshot!(t(b"{1, 2, 3, :soleil:}, ln"));
    assert_debug_snapshot!(t(b"add 1 :2:"));
    assert_debug_snapshot!(t(b":42:, add 1"));
    assert_debug_snapshot!(t(b"add 1, tostr"));
    assert_debug_snapshot!(t(b"add 1, tonum")); // TODO: incorrect if coercion should occure
    assert_debug_snapshot!(t(b"prout 1, caca"));
    assert_debug_snapshot!(t(b"_, split_, map_, join_"));
    assert_debug_snapshot!(t(b"{const, _, add, map}"));
    assert_debug_snapshot!(t(b"2 -"));
    assert_debug_snapshot!(t(b"- 2"));
    assert_debug_snapshot!(t(b"add 1 2 3"));
    assert_debug_snapshot!(t(b"3, add 1 2"));
    assert_debug_snapshot!(t(b":1:, tostr, _"));
    assert_debug_snapshot!(t(b"iterate [_ 1, add 1] 1")); // _ :: Num -> Num -> Num
    assert_debug_snapshot!(t(b"iterate [div 1, _ 1] 1")); // _ :: Num -> Num -> Num
    assert_debug_snapshot!(t(b"iterate [div 1, add 1] 1, _")); // _ :: [Num]+ -> returnof(typeof(_))
    assert_debug_snapshot!(t(b"let {a, b} [add a b] [panic::]"));
    assert_debug_snapshot!(t(b"{1, 2, 3}, let {h,, t} h [panic::]"));
    assert_debug_snapshot!(t(b"{1, 2, 3}, let {h,, t} t [panic::]"));
    assert_debug_snapshot!(t(b"repeat 1, let {h,, t} t [panic::]"));
    assert_debug_snapshot!(t(b"let a let b a b a")); // note: is syntax error
    assert_debug_snapshot!(t(b"let {a b, c} 0"));
    assert_debug_snapshot!(t(b"[1, let 0 fst snd] 1=:a:"));
    assert_debug_snapshot!(t(b"add 1, map, flip apply {1, 2, 3}"));
    // not quite a syntax error but only out of luck:
    // ```sel
    // [let a
    //     [[let b
    //         a
    //     ]                           # b -> a
    //         [panic: unreachable:]]  # a
    // ]                               # a -> a
    //     [panic: unreachable:]       # a
    // ```
    assert_debug_snapshot!(t(b"
let a
    [let b
        a
        [panic: unreachable:]]
    [panic: unreachable:]
"));
    assert_debug_snapshot!(t(b"use :bidoof: bdf, bdf-main"));
    assert_debug_snapshot!(t(b"
def head:
    returns the first item:
    [ let {h,, t}
        h
        [panic: head on empty list:]
    ],
head {1, 2, 3}
"));
    // TODO: not passing because.. uh.. 'paramof(sum)' doesn't get assigned correctly
    assert_debug_snapshot!(t(b"
def sum:: [
    let {h,, t}
        [sum t, add h]
        0
],
{sum, _}
"));
    // TODO: not passing because of src/parse.rs:1177
    assert_debug_snapshot!(t(b"
def a:: [b 2],
def b:: [add 1],
a
"));
}

#[test]
fn interpreting() {
    fn t(script: &[u8]) -> String {
        let mut global = Global::with_builtin();
        let source = global.registry.add_bytes("<test>", script.into());
        let result = parse::process(source, &mut global);
        let mut out = Vec::new();
        interp::run_write(&result.tree.unwrap(), &global, &mut out).unwrap();
        String::from_utf8(out).unwrap()
    }

    assert_snapshot!(t(b"0"));
    assert_snapshot!(t(b"add 1 2"));
    assert_snapshot!(t(b"let a [add a 1], flip apply 2"));
}
