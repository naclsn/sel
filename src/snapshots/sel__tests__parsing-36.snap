---
source: src/tests.rs
expression: "t(b\"\nlet a\n    [let b\n        a\n        [panic: unreachable:]]\n    [panic: unreachable:]\n\")"
---
Ok(
    (
        Named(
            "a=a",
        ),
        Tree {
            loc: Location(
                0,
                1..84,
            ),
            ty: 3,
            value: Apply(
                Bind(
                    Name(
                        Location(
                            0,
                            5..6,
                        ),
                        "a",
                    ),
                    Tree {
                        loc: Location(
                            0,
                            11..58,
                        ),
                        ty: 3,
                        value: Apply(
                            Bind(
                                Name(
                                    Location(
                                        0,
                                        16..17,
                                    ),
                                    "b",
                                ),
                                Tree {
                                    loc: Location(
                                        0,
                                        26..27,
                                    ),
                                    ty: 3,
                                    value: Apply(
                                        Name(
                                            "a",
                                        ),
                                        [],
                                    ),
                                },
                                Tree {
                                    loc: Location(
                                        0,
                                        0..0,
                                    ),
                                    ty: 0,
                                    value: Number(
                                        0.0,
                                    ),
                                },
                            ),
                            [
                                Tree {
                                    loc: Location(
                                        0,
                                        36..57,
                                    ),
                                    ty: 6,
                                    value: Apply(
                                        Name(
                                            "panic",
                                        ),
                                        [
                                            Tree {
                                                loc: Location(
                                                    0,
                                                    42..56,
                                                ),
                                                ty: 1,
                                                value: Bytes(
                                                    [
                                                        32,
                                                        117,
                                                        110,
                                                        114,
                                                        101,
                                                        97,
                                                        99,
                                                        104,
                                                        97,
                                                        98,
                                                        108,
                                                        101,
                                                    ],
                                                ),
                                            },
                                        ],
                                    ),
                                },
                            ],
                        ),
                    },
                    Tree {
                        loc: Location(
                            0,
                            0..0,
                        ),
                        ty: 0,
                        value: Number(
                            0.0,
                        ),
                    },
                ),
                [
                    Tree {
                        loc: Location(
                            0,
                            63..84,
                        ),
                        ty: 9,
                        value: Apply(
                            Name(
                                "panic",
                            ),
                            [
                                Tree {
                                    loc: Location(
                                        0,
                                        69..83,
                                    ),
                                    ty: 1,
                                    value: Bytes(
                                        [
                                            32,
                                            117,
                                            110,
                                            114,
                                            101,
                                            97,
                                            99,
                                            104,
                                            97,
                                            98,
                                            108,
                                            101,
                                        ],
                                    ),
                                },
                            ],
                        ),
                    },
                ],
            ),
        },
    ),
)
